File: src/tools/platform_generator_main.cpp
#include "platform/PlatformBuilder.hpp"
#include "platform/EdgePlatform.hpp"
#include "platform/FogPlatform.hpp"
#include "platform/CloudPlatform.hpp"
#include <iostream>
#include <string>

using namespace enigma;

void printUsage(const char* progName) {
    std::cout << "Usage: " << progName << " <type> [options]\n\n";
    std::cout << "Platform types:\n";
    std::cout << "  Simple (all hosts interconnected):\n";
    std::cout << "    edge <num_devices>            - Edge platform\n";
    std::cout << "    fog <num_nodes>               - Fog platform\n";
    std::cout << "    cloud <num_servers>           - Cloud platform\n";
    std::cout << "    iot <sensors> <actuators>     - IoT platform\n";
    std::cout << "\n  Cluster-based (organized in clusters):\n";
    std::cout << "    edge-cluster <num_clusters> <nodes_per_cluster>  - Edge clusters\n";
    std::cout << "    fog-cluster <num_clusters> <nodes_per_cluster>   - Fog clusters\n";
    std::cout << "    cloud-cluster <num_clusters> <nodes_per_cluster> - Cloud clusters\n";
    std::cout << "    hybrid-cluster <edge_clusters> <edge_nodes> <fog_clusters> <fog_nodes> <cloud_clusters> <cloud_nodes> [edge_cloud_direct] [output_file] - Flat hybrid (optional direct Edge-Cloud + optional output filename)\n";
    std::cout << "\nExamples:\n";
    std::cout << "  Simple:\n";
    std::cout << "    " << progName << " edge 10\n";
    std::cout << "\n  Clusters:\n";
    std::cout << "    " << progName << " edge-cluster 3 5       # 3 edge clusters with 5 devices each\n";
    std::cout << "    " << progName << " hybrid-cluster 2 10 2 5 1 20 1 custom_platform.xml  # Flat hybrid with direct Edge-Cloud links + custom file\n";
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        printUsage(argv[0]);
        return 1;
    }
    
    std::string type = argv[1];
    
    try {
        // Simple host-based platforms
        if (type == "edge" && argc >= 3) {
            int numDevices = std::stoi(argv[2]);
            std::cout << "Generating Edge platform with " << numDevices << " devices...\n";
            
            auto zone = EdgePlatform::createStarTopology(numDevices);
            PlatformGenerator gen;
            gen.generatePlatform("platforms/edge_platform.xml", zone);
            
        } else if (type == "fog" && argc >= 3) {
            int numNodes = std::stoi(argv[2]);
            std::cout << "Generating Fog platform with " << numNodes << " nodes...\n";
            
            auto zone = FogPlatform::createHierarchicalTopology(numNodes);
            PlatformGenerator gen;
            gen.generatePlatform("platforms/fog_platform.xml", zone);
            
        } else if (type == "cloud" && argc >= 3) {
            int numServers = std::stoi(argv[2]);
            std::cout << "Generating Cloud platform with " << numServers << " servers...\n";
            
            auto zone = CloudPlatform::createCluster(numServers);
            PlatformGenerator gen;
            gen.generatePlatform("platforms/cloud_platform.xml", zone);
            
        } else if (type == "hybrid") {
            std::cerr << "The 'hybrid' hierarchical mode is deprecated. Use 'hybrid-cluster-flat'.\n";
            return 1;
        } else if (type == "iot" && argc >= 4) {
            int sensors = std::stoi(argv[2]);
            int actuators = std::stoi(argv[3]);
            
            std::cout << "Generating IoT platform:\n";
            std::cout << "  - Sensors: " << sensors << "\n";
            std::cout << "  - Actuators: " << actuators << "\n";
            
            auto zone = EdgePlatform::createIoTPlatform(sensors, actuators);
            PlatformGenerator gen;
            gen.generatePlatform("platforms/iot_platform.xml", zone);
        
        // Cluster-based platforms
        } else if (type == "edge-cluster" && argc >= 4) {
            int numClusters = std::stoi(argv[2]);
            int nodesPerCluster = std::stoi(argv[3]);
            
            std::cout << "Generating Edge platform with " << numClusters 
                      << " clusters of " << nodesPerCluster << " nodes each...\n";
            
            std::vector<ClusterConfig> clusters;
            for (int i = 0; i < numClusters; i++) {
                clusters.emplace_back("edge_cluster_" + std::to_string(i), 
                                     nodesPerCluster, "1Gf", 1, "125MBps", "50us");
            }
            
            auto zone = PlatformGenerator::createEdgeWithClusters("edge_platform", clusters);
            PlatformGenerator gen;
            gen.generatePlatform("platforms/edge_platform.xml", zone);
            
        } else if (type == "fog-cluster" && argc >= 4) {
            int numClusters = std::stoi(argv[2]);
            int nodesPerCluster = std::stoi(argv[3]);
            
            std::cout << "Generating Fog platform with " << numClusters 
                      << " clusters of " << nodesPerCluster << " nodes each...\n";
            
            std::vector<ClusterConfig> clusters;
            for (int i = 0; i < numClusters; i++) {
                clusters.emplace_back("fog_cluster_" + std::to_string(i), 
                                     nodesPerCluster, "10Gf", 4, "1GBps", "10us");
            }
            
            auto zone = PlatformGenerator::createFogWithClusters("fog_platform", clusters);
            PlatformGenerator gen;
            gen.generatePlatform("platforms/fog_platform.xml", zone);
            
        } else if (type == "cloud-cluster" && argc >= 4) {
            int numClusters = std::stoi(argv[2]);
            int nodesPerCluster = std::stoi(argv[3]);
            
            std::cout << "Generating Cloud platform with " << numClusters 
                      << " clusters of " << nodesPerCluster << " nodes each...\n";
            
            std::vector<ClusterConfig> clusters;
            for (int i = 0; i < numClusters; i++) {
                clusters.emplace_back("cloud_cluster_" + std::to_string(i), 
                                     nodesPerCluster, "100Gf", 16, "10GBps", "1us");
            }
            
            auto zone = PlatformGenerator::createCloudWithClusters("cloud_platform", clusters);
            PlatformGenerator gen;
            gen.generatePlatform("platforms/cloud_platform.xml", zone);
            
        } else if (type == "hybrid-cluster" && argc >= 8) {
            int edgeClusters = std::stoi(argv[2]);
            int edgeNodes = std::stoi(argv[3]);
            int fogClusters = std::stoi(argv[4]);
            int fogNodes = std::stoi(argv[5]);
            int cloudClusters = std::stoi(argv[6]);
            int cloudNodes = std::stoi(argv[7]);
            bool directEdgeCloud = false;
            std::string outputFile = "platforms/hybrid_platform.xml";
            int optionalIndex = 8;
            if (argc > optionalIndex) {
                // Try parse edge_cloud_direct as integer; if not numeric treat as filename
                char* endPtr = nullptr;
                long val = strtol(argv[optionalIndex], &endPtr, 10);
                if (endPtr != nullptr && *endPtr == '\0') {
                    directEdgeCloud = (val != 0);
                    optionalIndex++;
                }
            }
            if (argc > optionalIndex) {
                outputFile = argv[optionalIndex];
                // if user did not include path, ensure it goes to platforms/ by default
                if (outputFile.find('/') == std::string::npos) {
                    outputFile = std::string("platforms/") + outputFile;
                }
            }
            
            std::cout << "Generating flat hybrid platform with clusters:\n";
            std::cout << "  - Edge: " << edgeClusters << " clusters × " << edgeNodes << " nodes\n";
            std::cout << "  - Fog: " << fogClusters << " clusters × " << fogNodes << " nodes\n";
            std::cout << "  - Cloud: " << cloudClusters << " clusters × " << cloudNodes << " nodes\n";
            std::cout << "  - Direct Edge-Cloud links: " << (directEdgeCloud ? "ENABLED" : "DISABLED") << "\n";
            std::cout << "  - Output file: " << outputFile << "\n";
            
            std::vector<ClusterConfig> edgeClustersVec, fogClustersVec, cloudClustersVec;
            
            for (int i = 0; i < edgeClusters; i++) {
                edgeClustersVec.emplace_back("edge_cluster_" + std::to_string(i), 
                                            edgeNodes, "1Gf", 1, "125MBps", "50us");
            }
            
            for (int i = 0; i < fogClusters; i++) {
                fogClustersVec.emplace_back("fog_cluster_" + std::to_string(i), 
                                           fogNodes, "10Gf", 4, "1GBps", "10us");
            }
            
            for (int i = 0; i < cloudClusters; i++) {
                cloudClustersVec.emplace_back("cloud_cluster_" + std::to_string(i), 
                                             cloudNodes, "100Gf", 16, "10GBps", "1us");
            }
            
            auto zone = PlatformGenerator::createHybridWithClustersFlat(edgeClustersVec, fogClustersVec, cloudClustersVec, directEdgeCloud);
            PlatformGenerator gen;
            gen.generatePlatform(outputFile, zone);
            
        } else {
            std::cerr << "Error: Invalid arguments\n\n";
            printUsage(argv[0]);
            return 1;
        }
        
        std::cout << "\nPlatform successfully generated!\n";
        
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }
    
    return 0;
}

File: src/python/tools/platform_generator_main.py
#!/usr/bin/env python3
"""
platform_generator_main.py – CLI tool for ENIGMA Python platform generation.

Mirrors src/tools/platform_generator_main.cpp.

Usage
-----
python platform_generator_main.py <type> [options]

Platform types
--------------
  Simple (all hosts interconnected):
    edge    <num_devices>                         - Edge star topology
    fog     <num_nodes>                           - Fog hierarchical topology
    cloud   <num_servers>                         - Cloud cluster
    iot     <sensors> <actuators>                 - IoT platform

  Cluster-based:
    edge-cluster   <num_clusters> <nodes_per_cluster>
    fog-cluster    <num_clusters> <nodes_per_cluster>
    cloud-cluster  <num_clusters> <nodes_per_cluster>
    hybrid-cluster <edge_clusters> <edge_nodes>
                   <fog_clusters>  <fog_nodes>
                   <cloud_clusters> <cloud_nodes>
                   [edge_cloud_direct] [output_file] [--generate-app]

Examples
--------
  python platform_generator_main.py edge 10
  python platform_generator_main.py edge-cluster 3 5
  python platform_generator_main.py hybrid-cluster 2 10 2 5 1 20
  python platform_generator_main.py hybrid-cluster 2 10 2 5 1 20 1 my_hybrid.xml --generate-app
"""
from __future__ import annotations

import argparse
import os
import sys
import textwrap

# Make sure the package on src/python is importable when run directly
_HERE = os.path.dirname(os.path.abspath(__file__))
_PYTHON_ROOT = os.path.join(_HERE, "..")
if _PYTHON_ROOT not in sys.path:
    sys.path.insert(0, _PYTHON_ROOT)

from enigma.platform import (
    ClusterConfig,
    PlatformGenerator,
    EdgePlatform,
    FogPlatform,
    CloudPlatform,
)


# --------------------------------------------------------------------------- #
# Template-app generator                                                       #
# --------------------------------------------------------------------------- #

def generate_template_app(
    app_filename: str,
    platform_file: str,
    edge_clusters: int, edge_nodes: int,
    fog_clusters: int,  fog_nodes: int,
    cloud_clusters: int, cloud_nodes: int,
) -> None:
    """Generate a Python template application skeleton for a given platform layout."""
    os.makedirs(os.path.dirname(os.path.abspath(app_filename)), exist_ok=True)

    lines: list[str] = [
        '"""',
        f'Template application for platform: {platform_file}',
        '',
        f'Configuration:',
        f'  Edge clusters : {edge_clusters} × {edge_nodes} nodes',
        f'  Fog  clusters : {fog_clusters} × {fog_nodes} nodes',
        f'  Cloud clusters: {cloud_clusters} × {cloud_nodes} nodes',
        '',
        'TODO: Implement actor logic in each class below.',
        '"""',
        '',
        'import sys',
        'import simgrid',
        '',
    ]

    if edge_clusters > 0:
        lines += [
            '',
            'class EdgeDevice:',
            '    """Edge device actor – sense, filter, forward."""',
            '',
            '    def __call__(self) -> None:',
            '        host = simgrid.this_actor.get_host()',
            '        simgrid.this_actor.info(f"[EDGE] Device \'{host.name}\' started")',
            '        # TODO: add edge processing logic here',
            '        # Example: collect data, execute locally, send to fog',
            '        simgrid.this_actor.info(f"[EDGE] Device \'{host.name}\' finished")',
            '',
        ]

    if fog_clusters > 0:
        lines += [
            '',
            'class FogNode:',
            '    """Fog node actor – aggregate, pre-process, forward."""',
            '',
            '    def __call__(self) -> None:',
            '        host = simgrid.this_actor.get_host()',
            '        simgrid.this_actor.info(f"[FOG] Node \'{host.name}\' started")',
            '        # TODO: add fog processing logic here',
            '        # Example: receive from edge, aggregate, filter, forward to cloud',
            '        simgrid.this_actor.info(f"[FOG] Node \'{host.name}\' finished")',
            '',
        ]

    if cloud_clusters > 0:
        lines += [
            '',
            'class CloudServer:',
            '    """Cloud server actor – heavy analytics, storage."""',
            '',
            '    def __call__(self) -> None:',
            '        host = simgrid.this_actor.get_host()',
            '        simgrid.this_actor.info(f"[CLOUD] Server \'{host.name}\' started")',
            '        # TODO: add cloud processing logic here',
            '        # Example: receive summaries, perform ML, store results',
            '        simgrid.this_actor.info(f"[CLOUD] Server \'{host.name}\' finished")',
            '',
        ]

    lines += [
        '',
        'def main() -> None:',
        '    e = simgrid.Engine(sys.argv)',
        '',
        '    if len(sys.argv) < 2:',
        '        print(f"Usage: {sys.argv[0]} <platform_file.xml>")',
        '        sys.exit(1)',
        '',
        '    e.load_platform(sys.argv[1])',
        '',
        '    hosts = e.all_hosts',
        '    simgrid.this_actor.info("=== Application Template ===")',
        '    simgrid.this_actor.info(f"Platform loaded with {len(hosts)} hosts")',
        '',
        '    edge_hosts  = [h for h in hosts if "edge"  in h.name]',
        '    fog_hosts   = [h for h in hosts if "fog"   in h.name]',
        '    cloud_hosts = [h for h in hosts if "cloud" in h.name]',
        '',
        '    simgrid.this_actor.info(',
        '        f"Detected: {len(edge_hosts)} edge, {len(fog_hosts)} fog, '
        '{len(cloud_hosts)} cloud hosts"',
        '    )',
        '',
    ]

    if edge_clusters > 0:
        lines += [
            '    for host in edge_hosts:',
            '        simgrid.Actor.create("edge_device", host, EdgeDevice())',
            '',
        ]
    if fog_clusters > 0:
        lines += [
            '    for host in fog_hosts:',
            '        simgrid.Actor.create("fog_node", host, FogNode())',
            '',
        ]
    if cloud_clusters > 0:
        lines += [
            '    for host in cloud_hosts:',
            '        simgrid.Actor.create("cloud_server", host, CloudServer())',
            '',
        ]

    lines += [
        '    e.run()',
        '    simgrid.this_actor.info("=== Simulation completed ===")',
        f'    simgrid.this_actor.info(f"Simulated time: {{simgrid.Engine.clock:.2f}} s")',
        '',
        '',
        'if __name__ == "__main__":',
        '    main()',
    ]

    with open(app_filename, "w", encoding="utf-8") as f:
        f.write("\n".join(lines) + "\n")

    print(f"Template application generated: {app_filename}")
    print("\nNext steps:")
    print(f"  1. Edit {app_filename} and implement your actor logic")
    print(f"  2. Run: python {app_filename} {platform_file}")


# --------------------------------------------------------------------------- #
# Main CLI                                                                     #
# --------------------------------------------------------------------------- #

def print_usage() -> None:
    print(textwrap.dedent(__doc__ or "").strip())


def main(argv: list[str] | None = None) -> int:  # noqa: C901
    argv = argv or sys.argv[1:]

    if not argv:
        print_usage()
        return 1

    platform_type = argv[0]
    gen = PlatformGenerator()

    try:
        # ------------------------------------------------------------------ #
        # Simple (host-based) platforms                                       #
        # ------------------------------------------------------------------ #
        if platform_type == "edge" and len(argv) >= 2:
            n = int(argv[1])
            print(f"Generating Edge platform with {n} devices…")
            zone = EdgePlatform.create_star_topology(n)
            gen.generate_platform("platforms/edge_platform.xml", zone)

        elif platform_type == "fog" and len(argv) >= 2:
            n = int(argv[1])
            print(f"Generating Fog platform with {n} nodes…")
            zone = FogPlatform.create_hierarchical_topology(n)
            gen.generate_platform("platforms/fog_platform.xml", zone)

        elif platform_type == "cloud" and len(argv) >= 2:
            n = int(argv[1])
            print(f"Generating Cloud cluster with {n} servers…")
            zone = CloudPlatform.create_cluster(n)
            gen.generate_platform("platforms/cloud_platform.xml", zone)

        elif platform_type == "iot" and len(argv) >= 3:
            sensors    = int(argv[1])
            actuators  = int(argv[2])
            print(f"Generating IoT platform: {sensors} sensors, {actuators} actuators…")
            zone = EdgePlatform.create_iot_platform(sensors, actuators)
            gen.generate_platform("platforms/iot_platform.xml", zone)

        # ------------------------------------------------------------------ #
        # Cluster-based platforms                                             #
        # ------------------------------------------------------------------ #
        elif platform_type == "edge-cluster" and len(argv) >= 3:
            num_clusters = int(argv[1])
            nodes        = int(argv[2])
            print(f"Generating Edge cluster platform: {num_clusters} clusters × {nodes} nodes…")
            clusters = [
                ClusterConfig(f"edge_cluster_{i}", nodes, "1Gf", 1, "125MBps", "50us")
                for i in range(num_clusters)
            ]
            zone = PlatformGenerator.create_edge_with_clusters("edge_platform", clusters)
            gen.generate_platform("platforms/edge_platform.xml", zone)

        elif platform_type == "fog-cluster" and len(argv) >= 3:
            num_clusters = int(argv[1])
            nodes        = int(argv[2])
            print(f"Generating Fog cluster platform: {num_clusters} clusters × {nodes} nodes…")
            clusters = [
                ClusterConfig(f"fog_cluster_{i}", nodes, "10Gf", 4, "1GBps", "10us")
                for i in range(num_clusters)
            ]
            zone = PlatformGenerator.create_fog_with_clusters("fog_platform", clusters)
            gen.generate_platform("platforms/fog_platform.xml", zone)

        elif platform_type == "cloud-cluster" and len(argv) >= 3:
            num_clusters = int(argv[1])
            nodes        = int(argv[2])
            print(f"Generating Cloud cluster platform: {num_clusters} clusters × {nodes} nodes…")
            clusters = [
                ClusterConfig(f"cloud_cluster_{i}", nodes, "100Gf", 16, "10GBps", "1us")
                for i in range(num_clusters)
            ]
            zone = PlatformGenerator.create_cloud_with_clusters("cloud_platform", clusters)
            gen.generate_platform("platforms/cloud_platform.xml", zone)

        elif platform_type == "hybrid-cluster" and len(argv) >= 7:
            edge_c = int(argv[1])
            edge_n = int(argv[2])
            fog_c  = int(argv[3])
            fog_n  = int(argv[4])
            cloud_c = int(argv[5])
            cloud_n = int(argv[6])

            # Parse optional trailing arguments
            edge_cloud_direct = False
            output_file = "platforms/hybrid_platform.xml"
            generate_app = False

            for arg in argv[7:]:
                if arg == "--generate-app":
                    generate_app = True
                elif arg.lstrip("-+").isdigit():
                    edge_cloud_direct = bool(int(arg))
                else:
                    output_file = arg if "/" in arg else f"platforms/{arg}"

            print("Generating flat hybrid platform:")
            print(f"  Edge  : {edge_c} clusters × {edge_n} nodes")
            print(f"  Fog   : {fog_c} clusters × {fog_n} nodes")
            print(f"  Cloud : {cloud_c} clusters × {cloud_n} nodes")
            print(f"  Direct Edge-Cloud: {'YES' if edge_cloud_direct else 'NO'}")
            print(f"  Output: {output_file}")

            edge_clusters = [
                ClusterConfig(f"edge_cluster_{i}", edge_n, "1Gf", 1, "125MBps", "50us")
                for i in range(edge_c)
            ]
            fog_clusters = [
                ClusterConfig(f"fog_cluster_{i}", fog_n, "10Gf", 4, "1GBps", "10us")
                for i in range(fog_c)
            ]
            cloud_clusters = [
                ClusterConfig(f"cloud_cluster_{i}", cloud_n, "100Gf", 16, "10GBps", "1us")
                for i in range(cloud_c)
            ]

            zone = PlatformGenerator.create_hybrid_with_clusters_flat(
                edge_clusters, fog_clusters, cloud_clusters, edge_cloud_direct
            )
            gen.generate_platform(output_file, zone)

            if generate_app:
                base = os.path.splitext(os.path.basename(output_file))[0]
                app_file = os.path.join("tests", f"{base}_app.py")
                generate_template_app(
                    app_file, output_file,
                    edge_c, edge_n,
                    fog_c, fog_n,
                    cloud_c, cloud_n,
                )

        else:
            print(f"Error: unrecognised command or missing arguments: {argv}", file=sys.stderr)
            print_usage()
            return 1

    except (ValueError, IndexError) as exc:
        print(f"Error: {exc}", file=sys.stderr)
        return 1

    return 0


if __name__ == "__main__":
    sys.exit(main())
